OMMatlab.m

```diff
@@ -696,18 +695,18 @@ function simulate(obj,resultfile,simflags)
                 buildmodelexpr=join(["buildModel(",cNames(1),")"]);
                 buildModelmsg=obj.sendExpression(buildmodelexpr);
                 %disp(buildModelmsg(:))
-                if(~isempty(buildModelmsg(1)))
+                if(~(buildModelmsg(1)=="")) %This is a changed code
                     obj.linearFlag=true;
                     obj.xmlfile=replace(fullfile(obj.mattempdir,char(buildModelmsg(2))),'\','/');
                     obj.linearquantitylist=[];
                     obj.linearinputs=strings(0,0);
                     obj.linearoutputs=strings(0,0);
                     obj.linearstates=strings(0,0);
-                    xmlparse(obj)
-                    result=getLinearMatrix(obj);
+                    xmlparse(obj)                    
                 else
-                    disp(omc.sendExpression("getErrorString()"));
+                    disp(obj.sendExpression("getErrorString()"));
                 end
+                result=getLinearMatrix(obj);

```

Comment:
That is exactly the point I am trying to make. Even if the buildModel() fails, there has to be a result anyway. I'm sure you definitely know that the function must output some sort of a result in all the conditional cases irrespective of their failure or success. Therefore, if I use `result = getLinearMatrix(obj)` in the `if` case and do not give any output of the function `linearize()` in the `else` case, that also would give me an error. In fact, it did give me an error saying that the output `result` of the function `linearize()` is unassigned, when the `if` case failed.

Reply:
@crazy734 can you send your model where i can reproduce the error, if buildModel fails there should not be result, we must get the error message on why it failed

Reply:
Create an OpenModelica model using this code first in OpenModelica.

`model Linearization_Test
output Real x;
input Real u = 4;
initial equation
x = 2;
equation
der(x) = x^2+u;
  annotation(
    experiment(StartTime = 0, StopTime = 10));
end Linearization_Test;`

Then execute this code in MATLAB.

`import OMMatlab.*()
omc = OMMatlab();
omc.ModelicaSystem('Linearization_Test.mo','Linearization_Test');
omc.simulate();
sol = omc.getSolutions();
sol1 = omc.getSolutions([sol]);
omc.setLinearizationOptions(["startTime = 0", "stopTime = 10"]);
States = omc.linearize()
sol2 = omc.getSolutions([sol]);`

This would result in an error stating that `Output argument "result" (and maybe others) not assigned during call to "OMMatlab/linearize".`

Reply:
@crazy734, thanks for the model, i will test it tomorrow

Reply:
@crazy734, I tested the model and i could reproduce the error, i have made some changes to the code, the latest version of OpenModelica generates linearized model with the default filename `"linearized_model.mo"`, Also now we can get proper error messages on why linearization failed something like below for your model

```
>> omc.linearize()
Building linearized Model failed: [C:/Users/arupa54/AppData/Local/Temp/tp28701588_a40a_4f35_ac85_47d35b3b1a28/linearized_model.mo:14:3-15:7:writable] Error: Variable nan not found in scope linearized_model.
```
Please test the new changes and see if it works

Reply:
Hello @arun3688. Thank you for the changes. I use OpenModelica version 1.14.1. Therefore the older code would be applicable for me. However, I still see that the output variable `result` is unassigned and it therefore produces the same error when run. Yes, this time the error shows up saying that the linearization has failed. However, it does not show the above-mentioned error. I am depicting the errors I have produced below. 

`Linearization failed: tempdir/linear_Linearization_Test.mo not found
Output argument "result" (and maybe others) not assigned during call to "OMMatlab/linearize".`

This means that the condition on line 698 `isfile(obj.linearfile)` is returning false and therefore the code execution jumps to line 724. However, even in this case, I am unable to reproduce the error that you have produced with your correction. Before this modification, I got the result states matrix to have at least `nan` values to identify a result for not working.

You have two versions of the code for both older and newer versions of Modelica. Therefore, I commented out the code on lines 687 and 688 for the newer version and ran the older version from lines 693 to 696. There is also an error in this version as well. In my case with the same model I have presented before, I see the `obj.linearfile` does not exist and therefore condition on 693 to 696 must be satisfied. However, since the `obj.linearfile` is empty, and the `isfile` command takes only text input, it also gives an error. I think it should be modified to `isempty(obj.linearfile)`. When I perform this modification, it still gives me the error of non-assignment of the function result `result`. 

It would be great if you could execute this and reproduce these errors yourself as well.

Reply:
@crazy734 , which matlab version you are using, I tested the code on `Matlab R2017b`, There were some new features related to strings and chars were added in this version, but i still based on your error message, i see that `isfile` is working so you got the error `Linearization failed: tempdir/linear_Linearization_Test.mo not found`  , and why you did not get the same message like me because, `linear_Linearization_Test.mo` does not exist and `obj.sendExpression("getErrorString()")`  returns empty string, can you go and check in the working directory "linearized_XXXX.mo" is created, i see you have some problem related with tempDirectory, because for every OMCSession a random temp directory i created, you can get the working directory using the API `getWorkDirectory`

```
>> omc= OMMatlab()
>> omc.ModelicaSystem("C:/juliatest/lineartest/linearizationtest.mo", "Linearization_Test")
>> omc.getWorkDirectory()
ans =
    'C:/Users/XXXXX/AppData/Local/Temp/tpd5ce77c0_e9d8_4eb8_a189_9a1b01c1f681'
```

Reply:
@arun3688 I am using 2020b, the latest version. However, `linear_Linearization_Test.mo` does exist. The only attribute of the `obj` object not getting assigned is the `obj.linearfile`. This is empty when I comment out the code for the latest version of OpenModelica and run the older parts of the code from lines 693 to 696. Therefore, the `isfile` function is not taking an empty array as input on the line 698 condition `isfile(obj.linearfile)`. To correct this itself I think you inserted the code from lines 693 to 696. However, I think that the condition for the `if` statement must be corrected from `~isempty(obj.linearfile)` to `isempty(obj.linearfile)`. When I perform this correction and execute the code, I still get the following error. 

`Building linearized Model failed: [tempdir/linear_Linearization_Test.mo:6:3-6:31:writable] Error: Variable nan not found in scope linear_Linearization__Test.
Error: Error occurred while flattening model linear_Linearization__Test

Output argument "result" (and maybe others) not assigned during call to "OMMatlab/linearize".

Error in Linearizer_Test (line 6)
States = omc.linearize()`

To remove even this final error, I, therefore, suggested in my previous edit, that the `result` variable should be assigned no matter what the conditional case, even if the matrices `A, B, C, D` turn out to be `nan`, with the error message of course, so that the function and subsequently the code would be executed completely.

Reply:
@crazy734, I downloaded `Matlab2020b` and tested the latest code , for me it is working fine with correct error message, see below

```
>> version

ans =

    '9.9.0.1570001 (R2020b) Update 4'

>> omc=OMMatlab()

omc = 

  OMMatlab with no properties.

>> omc.ModelicaSystem("C:/juliatest/lineartest/linearizationtest.mo", "Linearization_Test")

ans = 

    "C:/Users/arupa54/AppData/Local/Temp/tp57031f83_8f72_4c34_8485_0dd41ce329c2"

>> omc.linearize()
Building linearized Model failed: [C:/Users/arupa54/AppData/Local/Temp/tp57031f83_8f72_4c34_8485_0dd41ce329c2/linearized_model.mo:14:3-15:7:writable] Error: Variable nan not found in scope linearized_model.
```
and also to check whether the `isfile(obj.linearfile)` does not work if it empty, i  statically added the line before the condition check 

```
obj.linearfile = "";
if(isfile(obj.linearfile))
```
and i get correct error message

```
>> omc.linearize()
Linearization failed:  not found
```

Reply:
@arun3688 Ah okay, you're not assigning output of `linearize()` function to a variable so you don't get that error. However, if an output variable is not assigned in the function for the error cases, then the execution of these cases would give the aforementioned error of output variable not being assigned.

Reply:
@crazy734, the below message is fine, even if the output variable result is not assigned, but you are given a proper error message on why it is failing

```
>> result = omc.linearize()
Building linearized Model failed: [C:/Users/arupa54/AppData/Local/Temp/tpf46cff38_87ab_43df_bc80_a5e8faad91c6/linearized_model.mo:14:3-15:7:writable] Error: Variable nan not found in scope linearized_model.
Output argument "result" (and maybe others) not assigned during call to "OMMatlab/linearize"
```

Reply:
@arun3688 Yes that's true. Then that's it. I do not have any more issues with the code. Shall I close the PR?